python/felis/metadata.py

```diff
@@ -0,0 +1,509 @@
+# This file is part of felis.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+import logging
+from typing import IO, Any, Literal
+
+import sqlalchemy.schema as sqa_schema
+from sqlalchemy import (
+    CheckConstraint,
+    Column,
+    Constraint,
+    Engine,
+    ForeignKeyConstraint,
+    Index,
+    MetaData,
+    Numeric,
+    PrimaryKeyConstraint,
+    ResultProxy,
+    Table,
+    UniqueConstraint,
+    create_mock_engine,
+    make_url,
+    text,
+)
+from sqlalchemy.engine.mock import MockConnection
+from sqlalchemy.engine.url import URL
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.types import TypeEngine
+
+from felis.datamodel import Schema
+from felis.db._variants import make_variant_dict
+
+from . import datamodel as dm
+from .db import sqltypes
+from .types import FelisType
+
+logger = logging.getLogger(__name__)
+
+
+class InsertDump:
+    """An Insert Dumper for SQL statements which supports writing messages
+    to stdout or a file.
+    """
+
+    def __init__(self, file: IO[str] | None = None) -> None:
+        """Initialize the insert dumper.
+
+        Parameters
+        ----------
+        file : `TextIOBase` or None, optional
+            The file to write the SQL statements to. If None, the statements
+            will be written to stdout.
+        """
+        self.file = file
+        self.dialect: Any | None = None
+
+    def dump(self, sql: Any, *multiparams: Any, **params: Any) -> None:
+        """Dump the SQL statement to a file or stdout.
+
+        Parameters
+        ----------
+        sql : `Any`
+            The SQL statement to dump.
+
+        multiparams : `Any`
```

Comment:
This is existing code which was copied from a module that is being deleted, and I don't want to change the typing in the context of this ticket. It will be a very ugly type signature, and this is not a user-facing part of Felis.